### ISSUE TITLE
Getting timeout at end of downloading big file

### ISSUE THREAD

**Issue by Waynezhi** [robinrodricks/FluentFTP#1059]
**FTP Server Type:** Apache FtpServer

**Client Computer OS:** Ubuntu

**FluentFTP Version:** 42.*

**Framework:** .NET 6
Hello, me again.

I got another ftp problem, once the big file (800MB) downloaded, the program stopped responding (got hang for days), because it is waiting for the response of *DOWNLOAD* command.

**Logs :**

```

Response: 230 User logged in, proceed.
Command: FEAT
Status: Waiting for response to: FEAT
Response: 211-Extensions supported
Response: SIZE
Response: MDTM
Response: REST STREAM
Response: LANG en;zh-tw;ja;is
Response: MLST Size;Modify;Type;Perm
Response: AUTH SSL
Response: AUTH TLS
Response: MODE Z
Response: UTF8
Response: TVFS
Response: MD5
Response: MMD5
Response: MFMT
Response: 211 End
Status: Text encoding: System.Text.UTF8Encoding+UTF8EncodingSealed
Command: OPTS UTF8 ON
Status: Waiting for response to: OPTS UTF8 ON
Response: 200 Command OPTS okay.
Command: SYST
Status: Waiting for response to: SYST
Response: 215 UNIX Type: Apache FtpServer
Status: Listing parser set to: Machine
Command: PWD
Status: Waiting for response to: PWD
Response: 257 "/" is current directory.
> GetFileSize("***/****.txt.gz", -1)
Command: SIZE ***/****.txt.gz
Status: Waiting for response to: SIZE ***/****.txt.gz
Response: 502
> GetModifiedTime("***/****.txt.gz")
Command: MDTM ***/****.txt.gz
Status: Waiting for response to: MDTM ***/****.txt.gz
Response: 213 20221124162249.000
File size: -1.0 bytes, last modified: 11/24/2022 4:22:49 PM
> DownloadFile("/tmp/20221124201940657_****.txt.gz", "***/****.txt.gz", Resume, Retry)
> GetFileSize("***/****.txt.gz", -1)
Command: SIZE ***/****.txt.gz
Status: Waiting for response to: SIZE ***/****.txt.gz
Response: 502
> OpenRead("***/****.txt.gz", Binary, 0, -1)
Command: TYPE I
Status: Waiting for response to: TYPE I
Response: 200 Command TYPE okay.
> OpenPassiveDataStreamAsync(PASV, "RETR ***/****.txt.gz", 0)
Command: PASV
Status: Waiting for response to: PASV
Response: 227 Entering Passive Mode (35,207,2,62,205,53)
Status: Connecting to IP #1= 35.207.2.62:52533
Command: RETR ***/****.txt.gz
Status: Waiting for response to: RETR ***/****.txt.gz
Response: 150 File status okay; about to open data connection.
Status: Downloaded 831871613 bytes
Status: Disposing FtpSocketStream(data connection)
793.3 MB transferred
Status: Waiting for response to: *DOWNLOAD*

```

**Comment by FanDjango** [comment]
Hi, you again :-)

```
Command: SIZE ***/****.txt.gz
Status: Waiting for response to: SIZE ***/****.txt.gz
Response: 502
```

This looks suspicious, somehow...

**Comment by FanDjango** [comment]
I have today implemented an Apache FTP Server, Version 1.2.0. I cannot reproduce such strange behaviour. What version/how old is the server (probably Java embedded) that you are connecting to?

**Comment by FanDjango** [comment]
I am not happy about the server that you are connecting to. It seems to be pretty flaky.

**Comment by Waynezhi** [comment]
So cannot get file size from the server, I’ve noticed that it’s ok for small files, but not for big files.

> I am not happy about the server that you are connecting to. It seems to be pretty flaky. FtpWebRequest no problem?

I’ll test with FtpWebRequest.

**Comment by Waynezhi** [comment]
again, I got this issue not every time, seems sometimes the server does not return a response after transferring finished, for `Status: Waiting for response to: *DOWNLOAD*`, and I got a timeout exception:
```
Status:   Waiting for response to: *DOWNLOAD*
Status:   Testing connectivity using Socket.Poll()...
Status:   Disposing FtpSocketStream(data connection)
Status:   IOException for file /tmp/tmpCiCtRY.tmp : Unable to read data from the transport connection: Connection timed out.
Status:   Failed to download file.
Status:   Testing connectivity using Socket.Poll()...
Command:  QUIT
Status:   Waiting for response to: QUIT
Warning:  FtpClient.Disconnect(): Exception caught and discarded while closing control connection: System.IO.IOException: Unable to read data from the transport connection: Connection reset by peer.
 ---> System.Net.Sockets.SocketException (104): Connection reset by peer
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   --- End of inner exception stack trace ---
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at FluentFTP.FtpSocketStream.Read(Byte[] buffer, Int32 offset, Int32 count) in /var/collector/FluentFTP/Streams/FtpSocketStream.cs:line 477
   at FluentFTP.FtpSocketStream.ReadLine(Encoding encoding) in /var/collector/FluentFTP/Streams/FtpSocketStream.cs:line 542
   at FluentFTP.Client.BaseClient.BaseFtpClient.GetReplyInternal(String command, Boolean exhaustNoop, Int32 timeOut) in /var/collector/FluentFTP/Client/BaseClient/GetReply.cs:line 110
   at FluentFTP.Client.BaseClient.BaseFtpClient.GetReplyInternal(String command) in /var/collector/FluentFTP/Client/BaseClient/GetReply.cs:line 34
   at FluentFTP.Client.BaseClient.BaseFtpClient.FluentFTP.IInternalFtpClient.ExecuteInternal(String command) in /var/collector/FluentFTP/Client/BaseClient/Execute.cs:line 65
   at FluentFTP.FtpClient.Execute(String command) in /var/collector/FluentFTP/Client/SyncClient/Execute.cs:line 20
   at FluentFTP.FtpClient.Disconnect() in /var/collector/FluentFTP/Client/SyncClient/Disconnect.cs:line 17
Status:   Disposing FtpSocketStream(control connection)

# Dispose()
Status:   Disposing FtpClient object...
Status:   Disposing FtpSocketStream(control connection)
```
and with FtpWebRequest:
```
Unhandled exception. System.Net.WebException: The underlying connection was closed: An unexpected error occurred on a receive
   at System.Net.FtpWebRequest.SyncRequestCallback(Object obj)
   at System.Net.CommandStream.InvokeRequestCallback(Object obj)
   at System.Net.CommandStream.Abort(Exception e)
   at System.Net.CommandStream.CheckContinuePipeline()
   at System.Net.FtpWebRequest.DataStreamClosed(CloseExState closeState)
   at System.Net.FtpDataStream.System.Net.ICloseEx.CloseEx(CloseExState closeState)
   at System.Net.FtpDataStream.Dispose(Boolean disposing)
   at System.Net.FtpDataStream.AsyncReadCallback(IAsyncResult ar)
--- End of stack trace from previous location ---
   at System.Net.FtpDataStream.EndRead(IAsyncResult ar)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncTrimPromise`1.Complete(TInstance thisRef, Func`3 endMethod, IAsyncResult asyncResult, Boolean requiresSynchronization)
--- End of stack trace from previous location ---

**Comment by FanDjango** [comment]
Perhaps the control connection is no longer resposive - perhaps think about using NOOP as a test.

Not a big hope, but perhaps... https://github.com/robinrodricks/FluentFTP/wiki/Noop

Still, I think the main problem is that this server is totally unreliable concerning the response to the SIZE command. The code is Java, but sadly it does not welcome you with a version number, so there is hardly any idea how old it is. The one that I installed is of course the newest one: 1.2.0

**Comment by Waynezhi** [comment]
using NOOP seems a workaround, I will make more tests later for sure.
thanks a lot!

**Comment by FanDjango** [comment]
If you are using NOOP, I would be very thankful to see a log of how that process ends, if you would be so kind.

**Comment by Waynezhi** [comment]
sure, here you are:
```
Status: Downloaded 1968017269 bytes
Status: Disposing FtpSocketStream(data connection)
1.8 GB transferred, 0 KB/s
Status: Waiting for response to: *DOWNLOAD*
Status: Waiting - 8 seconds left
Status: Waiting - 7 seconds left
Status: Waiting - 6 seconds left
Status: Waiting - 5 seconds left
Status: Waiting - 4 seconds left
Status: Waiting - 3 seconds left
Status: Waiting - 2 seconds left
Status: Waiting - 1 seconds left
Status: Waiting - 0 seconds left
Status: GetReply(...) sequence: 226,200,200,200,200,200,200,200,200,200,200,200,200,200,200,200,200
Response: 200 Command NOOP okay.
Status: File Verification: PASS
> Dispose()
Status: Disposing FtpClient object...
Status: Disposing FtpSocketStream(control connection)

**Comment by FanDjango** [comment]
Thanks. That looks quite right. You can experiment to send less NOOPs, maybe not so often. Depends on server.

**Comment by FanDjango** [comment]
`200 Command NOOP okay` is also quite non-standard and needs to be included in FluentFTP. This server is a catastrophe, IMO.

**Comment by FanDjango** [comment]
I would like to change the branch I made for you to handle that correctly, I hope you can test it. Please standby.

**Comment by FanDjango** [comment]
I have updated the branch and invite you to test again

**Comment by FanDjango** [comment]
Ok, ready.

**Comment by Waynezhi** [comment]
Hi, 
Thank you, I changed to send NOOP every 90 seconds, here is the end of logs:
```

1.8 GB transferred, 0   KB/s
--
Status: Waiting for response to:   *DOWNLOAD*
Skipped: 200 Command NOOP okay.
Skipped: 200 Command NOOP okay.
Skipped: 200 Command NOOP okay.
Skipped: 200 Command NOOP okay.
Skipped: 200 Command NOOP okay.
Status: Waiting - 8 seconds left
Status: Waiting - 7 seconds left
Status: Waiting - 6 seconds left
Status: Waiting - 5 seconds left
Status: Waiting - 4 seconds left
Status: Waiting - 3 seconds left
Status: Waiting - 2 seconds left
Status: Waiting - 1 seconds left
Status: Waiting - 0 seconds left
Status: GetReply(...) sequence:   226,200,200,200,200,200
Response: 226 Transfer complete.
Status: File Verification: PASS
> Dispose()
Status: Disposing FtpClient object...
Status: Disposing FtpSocketStream(control   connection)

**Comment by FanDjango** [comment]
That looks good now. You can continue using that version - soon I will put in the changes to the regular stream and eventually it will get released.

**Comment by Waynezhi** [comment]
Great, please let me know when the new release gets out.
Thanks again !

**Comment by FanDjango** [comment]
Released as V42.1.0
